pkg/storage/storagewrappers/boundedconcurrency.go

```diff
@@ -0,0 +1,86 @@
+package storagewrappers
+
+import (
+	"context"
+	"time"
+
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const TimeWaitingSpanAttribute = "time_waiting"
+
+var _ storage.RelationshipTupleReader = (*boundedConcurrencyTupleReader)(nil)
+
+var (
+	timeWaitingCounter = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name:    "time_waiting_for_bounded_query",
+		Help:    "Time (in ms) spent waiting for Read, ReadUserTuple and ReadUsersetTuples calls to the datastore",
+		Buckets: []float64{10, 25, 50, 100, 1000, 5000}, // milliseconds. Upper bound is config.UpstreamTimeout
+	})
+)
+
+type boundedConcurrencyTupleReader struct {
+	storage.RelationshipTupleReader
+	limiter chan struct{}
+}
+
+// NewBoundedConcurrencyTupleReader returns a wrapper over a datastore that makes sure that there are, at most,
+// N concurrent calls to Read, ReadUserTuple and ReadUsersetTuples.
+// Consumers can then rest assured that one client will not hoard all the database connections available.
+func NewBoundedConcurrencyTupleReader(wrapped storage.RelationshipTupleReader, N uint32) *boundedConcurrencyTupleReader {
+	return &boundedConcurrencyTupleReader{
+		RelationshipTupleReader: wrapped,
+		limiter:                 make(chan struct{}, N),
+	}
+}
+
+func (b *boundedConcurrencyTupleReader) ReadUserTuple(ctx context.Context, store string, tupleKey *openfgapb.TupleKey) (*openfgapb.Tuple, error) {
+	b.waitForLimiter(ctx)
+
+	defer func() {
+		<-b.limiter
+	}()
+
+	return b.RelationshipTupleReader.ReadUserTuple(ctx, store, tupleKey)
+}
+
+func (b *boundedConcurrencyTupleReader) Read(ctx context.Context, store string, tupleKey *openfgapb.TupleKey) (storage.TupleIterator, error) {
+	b.waitForLimiter(ctx)
+
+	defer func() {
+		<-b.limiter
+	}()
+
+	return b.RelationshipTupleReader.Read(ctx, store, tupleKey)
+}
+
+func (b *boundedConcurrencyTupleReader) ReadUsersetTuples(ctx context.Context, store string, filter storage.ReadUsersetTuplesFilter) (storage.TupleIterator, error) {
+	b.waitForLimiter(ctx)
+
+	defer func() {
+		<-b.limiter
+	}()
+
+	return b.RelationshipTupleReader.ReadUsersetTuples(ctx, store, filter)
+}
+
+func (b *boundedConcurrencyTupleReader) Close() {
+	close(b.limiter)
```

Comment:
You need to call `Close` on the underlying (e.g. wrapped) `RelationshipTupleReader`.

```go
b.RelationshipTupleReader.Close()
close(b.limiter)
```

Reply:
There is no method `Close()` on `RelationshipTupleReader`. Also, in this PR, we instantiate a new `boundedConcurrencyTupleReader` on every ListObjects request. I wouldn't want to close the underlying datastore on every request